src/Storages/MergeTree/KeyCondition.cpp

```diff
@@ -1292,8 +1292,8 @@ bool KeyCondition::tryParseAtomFromAST(const ASTPtr & node, ContextPtr context,
                 key_expr_type_not_null = key_expr_type;
 
             bool cast_not_needed = is_set_const /// Set args are already casted inside Set::createFromAST
-                || ((isNativeNumber(key_expr_type_not_null) || isDateTime(key_expr_type_not_null))
-                    && (isNativeNumber(const_type) || isDateTime(const_type))); /// Numbers and DateTime are accurately compared without cast.
+                || ((isInteger(key_expr_type_not_null) || isDateTime(key_expr_type_not_null))
+                    && (isInteger(const_type) || isDateTime(const_type))); /// Integers and DateTime are accurately compared without cast.
```

Comment:
Without cast for floats for conditions like `float_column > 0` we could get wrong results, because we will create integer range `(0, +inf)` for analysis and it could be optimized to [1, +inf) [here](https://github.com/ClickHouse/ClickHouse/blob/c347438612f70def27a7675bb13faa9ff1de98e8/src/Storages/MergeTree/KeyCondition.h#L103-L135), which is obviously wrong for floats in `(0, 1)`.